source/slang/slang-compiler.h

```diff
@@ -2185,40 +2186,257 @@ namespace Slang
     {
+        // TODO: The existence of this type is evidence of a design/architecture problem.
+        //
+        // A better formulation of things requires a few key changes:
+        //
+        // 1. All optional capabilities need to be enumerated as part of the `CapabilitySet`
+        //  system, so that they can be reasoned about uniformly across different targets
+        //  and different layers of the compiler.
+        //
+        // 2. The front-end should be responsible for either or both of:
+        //
+        //   * Checking that `public` or otherwise externally-visible items (declarations/definitions)
+        //     explicitly declare the capabilities they require, and that they only ever
+        //     make use of items that are comatible with those required capabilities.
+        //
+        //   * Inferring the capabilities required by items that are not externally visible,
+        //     and attaching those capabilities explicit as a modifier or other synthesized AST node.
+        //
+        // 3. The capabilities required by a given `ComponentType` and its entry points should be
+        // explicitly know-able, and they should be something we can compare to the capabilities
+        // of a code generation target *before* back-end code generation is started. We should be
+        // able to issue error messages around lacking capabilities in a way the user can understand,
+        // in terms of the high-level-language entities.
+
     public:
-        BackEndCompileRequest(
-            Linkage*        linkage,
-            DiagnosticSink* sink,
-            ComponentType*  program = nullptr);
+    };
 
-        // Should we dump intermediate results along the way, for debugging?
-        bool shouldDumpIntermediates = false;
+        /// A context for code generation in the compiler back-end
+    struct CodeGenContext
+    {
+    public:
+        typedef List<Index> EntryPointIndices;
 
-        ComponentType* getProgram() { return m_program; }
-        void setProgram(ComponentType* program) { m_program = program; }
+        struct Shared
+        {
+        public:
+            Shared(
+                TargetProgram*              targetProgram,
+                EntryPointIndices const&    entryPointIndices,
+                DiagnosticSink*             sink,
+                EndToEndCompileRequest*     endToEndReq)
+                : targetProgram(targetProgram)
+                , entryPointIndices(entryPointIndices)
+                , sink(sink)
+                , endToEndReq(endToEndReq)
+            {}
+
+//            Shared(
+//                TargetProgram*              targetProgram,
+//                EndToEndCompileRequest*     endToEndReq);
+
+            TargetProgram*          targetProgram = nullptr;
+            EntryPointIndices       entryPointIndices;
+            DiagnosticSink*         sink = nullptr;
+            EndToEndCompileRequest* endToEndReq = nullptr;
```

Comment:
What are the situations where we need the back reference to `EndToEndCompileRequest` here? Is it only for pass through? Is it possible to get rid of this to make things cleaner?

Reply:
A search for `isEndToEndCompile` in this PR reveals that it is really only needed to get dump ir options. Seems like this can just be cleaned up with the `IConfig` refactor we talked about earlier.

Reply:
The `endToEndReq` is really used in two main cases:

1. Pass-through compilation, in which case it is the only piece that has all the relevant state (since otherwise we cannot access the front-end configuration from within codegen).
2. Writing of output files for command-line compilation, in which case the end-to-end request is the only object that knows about the output paths.

It is likely we can split out the logic for (2) so that most of code generation doesn't need to care.

Pass-through (1) is a much more complicated case, though, and it will probably take a lot of thought to design how it ought to work more properly. I am worried that we are getting to a point where we may need to treat pass-through compilation as an entirely separate mode, rather than trying to integrate it into the ordinary functioning of the compiler as we have so far done.